File: scala-server/app/com/bolour/eagerwords/server/util/WordUtil.scala
/*
 * Copyright 2017-2020 Azad Bolour
 * Licensed under GNU Affero General Public License v3.0 -
 *    https://github.com/azadbolour/eagerwords.com/blob/master/LICENSE.md
 */
package com.bolour.eagerwords.server.util

import org.slf4j.LoggerFactory

import scala.collection.mutable
import scala.collection.mutable.HashSet

object WordUtil {
  val logger = LoggerFactory.getLogger(this.getClass)

  type Length = Int             // length of board strip, word, etc.
  type NumBlanks = Int          // number of blanks in a board strip
  type LetterCombo = String     // combination of letters (sorted with dups)
  type DictWord = String        // dictionary word
  type WordsByCombo = Map[LetterCombo, List[DictWord]] // words by letter combination

  val english = "en"
  val defaultLanguageCode = english

  val blankChar: Char = ' '
  def isBlankChar(ch: Char): Boolean = ch == blankChar
  val blackChar: Char = '-'
  def isBlackChar(ch: Char): Boolean = ch == blackChar

  def stringToLetterCombo(s: String): LetterCombo = s.sorted

  def computeCombosGroupedByLength(letters: String): Map[Length, List[LetterCombo]] = {
    computeCombos(letters).groupBy(_.length)
  }

  def computeCombos(letters: String): List[LetterCombo] = {
    // remove the empty combo (head) and convert elements to sort strings
    computeCombosUnsorted(letters.toList).tail.map(_.sorted.mkString)
  }

  /** includes the empty combo as the first element */
  private def computeCombosUnsorted(letters: List[Char]): List[List[Char]] = {
    letters match {
      case Nil => List(Nil)
      case head :: tail =>
        val tailCombos = computeCombosUnsorted(tail)
        val headCombos = tailCombos.map(head :: _)
        tailCombos ++ headCombos
    }
  }

  def mergeLetterCombos(combo1: LetterCombo, combo2: LetterCombo): LetterCombo =
    stringToLetterCombo(combo1.concat(combo2))

  /**
    * Compute the set of all versions of a string masked by replacing
    * 0 to n of its characters with blanks.
    *
    * @param s The string.
    * @param n Maximum number of characters to mask.
    * @return Masked versions of the string.
    */
  def maskWithBlanks(s: String, n: Int): List[String] = {
    def prepend(ch: Char, set: List[String]) = set map { ch +: _ }
    val maskedSet =
      if (n == 0)
        List(s)
      else if (s.length == 1)
        List(blankChar.toString, s) // n > 0 but we can have at most 1 blank in this case, i.e., 0 or 1.
      else
        prepend(s.head, maskWithBlanks(s.tail, n)) ++ prepend(blankChar, maskWithBlanks(s.tail, n - 1))
    maskedSet
  }
}

File: haskell-server/src/Bolour/Language/Util/WordUtil.hs
--
-- Copyright 2017-2020 Azad Bolour
-- Licensed under GNU Affero General Public License v3.0 -
--   https://github.com/azadbolour/eagerwords/blob/master/LICENSE.md
--

module Bolour.Language.Util.WordUtil (
    LetterCombo
  , DictWord
  , WordIndex
  , ByteCount
  , BlankCount
  , mkLetterCombo
  , mergeLetterCombos
  , computeCombos
  , computeCombosGroupedByLength
  , maskWithBlanks
  , blankChar, isBlankChar, blackChar, isBlackChar
  ) where

-- import Data.ByteString.Char8
-- import qualified Data.ByteString.Char8 as BS
import Data.Map (Map)
import Data.List (sort)
import qualified Data.Map as Map
import qualified Data.Maybe as Maybe

import qualified Bolour.Util.MiscUtil as MiscUtil

-- | Combinations of letters (with repetition) sorted in a byte string.
type LetterCombo = String

-- | A dictionary word.
type DictWord = String

-- | Index of words.
--   Key is a combination of letters.
--   Value is permutations of the letters in the key that are actual words.
type WordIndex = Map LetterCombo [DictWord]

type ByteCount = Int
type BlankCount = Int

blackChar = '-'
isBlackChar :: Char -> Bool
isBlackChar = (== blackChar)

blankChar = ' '
isBlankChar :: Char -> Bool
isBlankChar = (== blankChar)

-- | Make a permutation of some letters, create the corresponding combination of those letters.
--   In our representation just sort the letters.
mkLetterCombo :: String -> LetterCombo
mkLetterCombo permutation = sort permutation

-- | Merge two combinations of letters.
mergeLetterCombos :: LetterCombo -> LetterCombo -> LetterCombo
mergeLetterCombos combo1 combo2 = sort $ combo1 ++ combo2

-- -- | Look up the words that are permutations of a given combination of letters.
-- lookupWordIndex :: LetterCombo -> WordIndex -> [DictWord]
-- lookupWordIndex letters wordIndex = Maybe.fromMaybe [] (Map.lookup letters wordIndex)

-- TODO. Eliminate duplicates for maximum performance.

-- | Compute all combinations of a set of letters and group them by length.
computeCombosGroupedByLength :: String -> Map ByteCount [LetterCombo]
computeCombosGroupedByLength bytes =
  let combos = computeCombos bytes
  in MiscUtil.mapFromValueList length combos

computeCombos :: String -> [LetterCombo]
computeCombos bytes = sort <$> computeCombosUnsorted bytes

computeCombosUnsorted :: String -> [String]
computeCombosUnsorted bytes
  | null bytes = [""]
  | otherwise =
      let h = head bytes
          t = tail bytes
          tailCombos = computeCombosUnsorted t
          combosWithHead = (:) h <$> tailCombos
      in tailCombos ++ combosWithHead

maskWithBlanks :: String -> Int -> [String]
maskWithBlanks s maxBlanks
  | maxBlanks == 0 = [s]
  | length s == 1 = [[blankChar], s]
  | otherwise = let head:tail = s
                in prepend head (maskWithBlanks tail maxBlanks) ++ prepend blankChar (maskWithBlanks tail (maxBlanks - 1))

prepend :: Char -> [String] -> [String]
prepend ch ss = (\s -> ch:s) <$> ss -- TODO. avoid lambda - can use section: (ch :) <$> ss